src/FSharp.Control.TaskSeq.Test/TaskSeq.TakeWhile.Tests.fs

```diff
@@ -0,0 +1,219 @@
+
+module SideEffects =
+    [<Theory; ClassData(typeof<TestSideEffectTaskSeq>)>]
+    let ``TaskSeq-takeWhile filters correctly`` variant =
+        Gen.getSeqWithSideEffect variant
+        |> TaskSeq.takeWhile condWithGuard
+        |> TaskSeq.map char
+        |> TaskSeq.map ((+) '@')
+        |> TaskSeq.toArrayAsync
+        |> Task.map (String >> should equal "ABCDE")
+
+    [<Theory; ClassData(typeof<TestSideEffectTaskSeq>)>]
+    let ``TaskSeq-takeWhileAsync filters correctly`` variant =
+        Gen.getSeqWithSideEffect variant
+        |> TaskSeq.takeWhileAsync (fun x -> task { return condWithGuard x })
+        |> TaskSeq.map char
+        |> TaskSeq.map ((+) '@')
+        |> TaskSeq.toArrayAsync
+        |> Task.map (String >> should equal "ABCDE")
+
+    [<Theory; InlineData(false, false); InlineData(true, false); InlineData(false, true); InlineData(true, true)>]
+    let ``TaskSeq-takeWhile(Inclusive)?(Async)? __special-case__ prove it does not read beyond the failing yield`` (inclusive, async) = task {
+        let mutable x = 42 // for this test, the potential mutation should not actually occur
+
+        let items = taskSeq {
+            yield x // Always passes the test; always returned
+            yield x * 2 // the failing item (which will also be yielded in the result when using *Inclusive)
+            x <- x + 1 // we are proving we never get here
+        }
+
+        let f =
+            match inclusive, async with
+            | false, false -> TaskSeq.takeWhile (fun x -> x = 42)
+            | true, false -> TaskSeq.takeWhileInclusive (fun x -> x = 42)
+            | false, true -> TaskSeq.takeWhileAsync (fun x -> task { return x = 42 })
+            | true, true -> TaskSeq.takeWhileInclusiveAsync (fun x -> task { return x = 42 })
+
+        let expected = if inclusive then [| 42; 84 |] else [| 42 |]
+
+        let! first = items |> f |> TaskSeq.toArrayAsync
+        let! repeat = items |> f |> TaskSeq.toArrayAsync
+
+        first |> should equal expected
+        repeat |> should equal expected
+        x |> should equal 42
+    }
+
+    [<Theory; InlineData(false, false); InlineData(true, false); InlineData(false, true); InlineData(true, true)>]
+    let ``TaskSeq-takeWhile(Inclusive)?(Async)? __special-case__ prove side effects are executed`` (inclusive, async) = task {
+        let mutable x = 41
+
+        let items = taskSeq {
+            x <- x + 1
+            yield x
+            x <- x + 2
+            yield x * 2
+            x <- x + 200 // as previously proven, we should not trigger this
+        }
+
+        let f =
+            match inclusive, async with
+            | false, false -> TaskSeq.takeWhile (fun x -> x < 50)
+            | true, false -> TaskSeq.takeWhileInclusive (fun x -> x < 50)
+            | false, true -> TaskSeq.takeWhileAsync (fun x -> task { return x < 50 })
+            | true, true -> TaskSeq.takeWhileInclusiveAsync (fun x -> task { return x < 50 })
```

Comment:
Creative! I like it :). I may generalize this soon(ish) for the other tests as well.

Reply:
Done.